src/DIRAC/Resources/Computing/BatchSystems/Host.py

```diff
@@ -195,7 +195,7 @@ def getCEStatus(self, **kwargs):
       infoFile.close()
       jobInfo = json.loads(jobInfo)
       pid = jobInfo['PID']
-      cmd = 'ps -f -p %s --no-headers | wc -l' % pid
+      cmd = 'bash -c "ps -f -p %s --no-headers | wc -l"' % pid
```

Comment:
If you want to count the lines use `len()` in Python instead of launching two extra processes and introducing a possible arbitrary code execution vulnerability.

Or just use `psutil` that will be faster and more robust.

Reply:
This code is executed by native python on a node without any special DIRAC environment. So, psutil apriori is not available.

Reply:
I tried to consider your suggestion and this looks ugly. "ps" gives non-zero status code if the process is not found but also it can be for some other reasons, so analyzing the returncode is not relevant then. Filtering the output through "wc -l" does not seem to be heavy, this was working for ages.

Reply:
Sorry I had read to quickly when checking.

Looking more closely this is still broken as adding the `bash -c` just makes the problem even worse as `shlex.split(cmd)`. I think you're looking for the `shell=True` argument to `Popen` but we shouldn't use that.

1. Am I right in thinking that checking the status code is a waste of time as it checks the return code of `wc -l` which will only fail if `wc` isn't installed on the host?

2. If so, how about this instead:

```python
sp = subprocess.Popen(["ps", "-f", "-p", str(pid), "--no-headers"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
output, error = sp.communicate()
if output:
  running += 1
  usedCores += jobInfo['NCores']
else:
  stamp = jobInfo['JOBID']
  jobLife = datetime.utcnow() - datetime.strptime(jobInfo['SubmissionTime'], "%Y-%m-%d %H:%M:%S")
  if jobLife > CLEAN_DELAY:
    self.__cleanJob(stamp, infoDir, workDir, outputDir, errorDir)
```

Reply:
The fix I did I tested in a production environment where I have suffered from this problem. Your suggestion can work also, I can try. But "ps" can give some errors on some nodes which in your case will be interpreted as detecting a running process (non empty output). So, I would prefer still to filter it through wc -l and check that the output is '1'.
The returncode checking does not make much sense indeed but it is always good to do. This is how I discovered that Popen without shell does not work.

Reply:
shlex is splitting the line as:

    $  shlex.split('bash -c "ps -f -p 111 --no-headers | wc -l"')
    ['bash', '-c', 'ps -f -p 111 --no-headers | wc -l']

And apparently it works with Popen.

Reply:
OK, did it without shell.